### PR TITLE
planner: support push predicates down to the memory table

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

### What problem does this PR solve? 

In the current implementation of the memory table which retrieves remote data, we use an inefficient way to retrieve remote data.

e.g:
```sql
SELECT * FROM tidb_cluster_config WHERE type='tikv' AND address='192.168.1.9:2379'
```
We must request all components in the cluster via HTTP API for retrieving
configruation and filter them by `type/address` columns.

### What is changed and how it works?

Define a `MemTablePredicateExtractor` to support push predicates down to the memory table.

The purpose of defining a `MemTablePredicateExtractor` is to optimize this scenario (especially for the huge cluster which contains much of nodes)
1. Define a `ClusterConfigTablePredicateExtractor`
2. Extract the `type/address` columns on the logic optimizing stage and save them via fields.
3. Passing the extractor to the `ClusterConfigReaderExec` executor
4. Executor sends requests to the target components instead of all of the components

**NOTE**

This PR is nearly 500 lines of change, so the `ClusterConfigReader` executor will be implemented in the next PR(https://github.com/pingcap/tidb/pull/13832).

### Check List 

Tests 

 - Unit test

Release note

 - No need

### CODE REVIEW

File: planner/core/memtable_predicate_extractor.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	// We should use INTERSECTION of sets because of the predicates is CNF array
+	for _, expr := range predicates {
+		var extracted bool
+		switch x := expr.(type) {
+		case *expression.ScalarFunction:
+			switch x.FuncName.L {
+			case ast.EQ:
+				colName, datum := e.extractColEqConsExpr(extractCols, x)
+				switch colName {
+				case ColNameType:
```
Comment (lamxTyler) [inline planner/core/memtable_predicate_extractor.go]: These four cases are almost the same, why not use a function for them?

File: planner/core/memtable_predicate_extractor.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+					tp := datum.GetString()
```
Comment (lamxTyler) [inline planner/core/memtable_predicate_extractor.go]: Is `tikv` and `TiKV` the same?

File: planner/core/memtable_predicate_extractor.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+					e.Addresses.Insert(addr)
+				}
+				extracted = colName != ""
+			case ast.In:
+				colName, datums := e.extractColInConsExpr(extractCols, x)
+				switch colName {
+				case ColNameType:
+					tmpNodeTypes := set.NewStringSet()
+					for _, datum := range datums {
+						tmpNodeTypes.Insert(datum.GetString())
```
Comment (lzmhhh123) [inline planner/core/memtable_predicate_extractor.go]: Can we assert the datum type is always string?

File: planner/core/memtable_predicate_extractor.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	e.NodeTypes = set.NewStringSet()
+	e.Addresses = set.NewStringSet()
+
+	// We should use INTERSECTION of sets because of the predicates is CNF array
+	for _, expr := range predicates {
+		var extracted bool
+		switch x := expr.(type) {
+		case *expression.ScalarFunction:
+			switch x.FuncName.L {
+			case ast.EQ:
```
Comment (lzmhhh123) [inline planner/core/memtable_predicate_extractor.go]: We can combine the two branches like this:
```go
var extractFunc func(extractCols map[int64]*types.FieldName, expr *expression.ScalarFunction) (string, []types.Datum)
switch x.FuncName.L {
case ast.EQ:
	extractFunc = e.extractColEqConsExpr
case ast.In
	extractFunc = e.extractColInConsExpr
}
colName, datums := extractFunc(extractCols, x)
// then deal the datums and colName.
```

File: planner/core/memtable_predicate_extractor.go (hunk 2 above)
Comment (lonng) [reply planner/core/memtable_predicate_extractor.go]: Nice catch.

File: planner/core/memtable_predicate_extractor.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	col, isCol := args[0].(*expression.Column)
+	if !isCol {
+		return "", types.Datum{}
+	}
+	name, found := extractCols[col.UniqueID]
+	if !found {
+		return "", types.Datum{}
+	}
+	// The `rhs` of EQ expression must be a constant
+	// SELECT * FROM t1 WHERE c = 'rhs'
```
[the remark this answers is not in the document]
Comment (lonng) [reply planner/core/memtable_predicate_extractor.go]: Nice catch, addressed

File: planner/core/memtable_predicate_extractor.go (hunk 1 above)
Comment (lonng) [reply planner/core/memtable_predicate_extractor.go]: done

File: planner/core/memtable_predicate_extractor.go (hunk 3 above)
Comment (lonng) [reply planner/core/memtable_predicate_extractor.go]: All datum can be converted to a string (node type and address always be a string here).

File: planner/core/memtable_predicate_extractor.go (hunk 4 above)
Comment (lonng) [reply planner/core/memtable_predicate_extractor.go]: done

File: planner/core/memtable_predicate_extractor.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	for i, name := range names {
+		if ln := name.ColName.L; ln == ColNameType || ln == ColNameAddress {
+			extractCols[schema.Columns[i].UniqueID] = name
+		}
+	}
+	// We don't support change the schema of memory table, so something must
+	// be wrong if the `type/address` columns cannot be found in the schema.
+	// The purpose of the following assert is used to make test failure because
+	// it will never happen if our release passing all test cases.
+	if len(extractCols) != 2 {
```
Comment (eurekaka) [inline planner/core/memtable_predicate_extractor.go]: I still don't understand the purpose of this check.

File: planner/core/memtable_predicate_extractor.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+			s = strings.ToLower(datum.GetString())
+		} else {
+			s = datum.GetString()
+		}
+		tmpNodeTypes.Insert(s)
+	}
+	if len(lhs) > 0 {
+		return lhs.Intersection(tmpNodeTypes)
+	}
+	return tmpNodeTypes
```
Comment (eurekaka) [inline planner/core/memtable_predicate_extractor.go]: It is pretty strange that intersection of empty set and another set is not empty.

**Comment by lonng** [comment]
> I looked into the current codebase and think that maybe `ExtractEqAndInCondition` is suitable for your case.

I investigated the function `ExtractEqAndInCondition` and it can be used to extract the expressions `EQ` and `IN`, but still some customized logic to extract the predicates because I need the predicates values instead of a range (prefer to keep current implementation and finding a better way to refactor this part).

File: planner/core/memtable_predicate_extractor.go (hunk 6 above)
Comment (lonng) [reply planner/core/memtable_predicate_extractor.go]: I have reworked this comment, PTAL again.

**Pull request merged**